### DISCUSSION TITLE
Block reads (and writes)

### DISCUSSION THREAD

**Discussion by piperoc** [open62541pp/open62541pp#279] [Q&A]
In [OPC UA Part 4 etc.](https://reference.opcfoundation.org/Core/Part4/v104/docs/5.10.2) the read requests contain a array of `NodesToRead` and `Results` to minimize the traffic on read operations (or subscriptions). 

I'm rather fond of the DataSource backends, although that would call a separate read per each node.

I have the feeling I would have to go down to the C API to implement my own block operations. 

Thanks you for sharing your thoughts.

**Comment by lukasberbuer** [comment]
You can bundle multiple read operations in a single `ReadRequest` from the client-side to minimize the traffic:
https://github.com/open62541pp/open62541pp/blob/v0.12.0/include/open62541pp/services/Attribute.h#L48-L60

But you don't have to prepare for this on the server-side. This is handled automatically inside the server.

Currently, the generic requests are only available as free functions in the `services` namespace. Mainly because this is a client-only implementation and the `Node` class shares the same interface for client and server.
Bundling multiple read operations server-side has no benefits because they are handled locally without any network transmission.
